keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/ArtifactInterfaces.kt

```diff
@@ -0,0 +1,82 @@
+package com.netflix.spinnaker.keel.api
+
+import com.netflix.spinnaker.keel.api.artifacts.ArtifactType
+
+/**
+ * Implemented by [ResourceSpec] or concrete resource types that (may) contain artifact information, typically compute
+ * resources.
+ *
+ * The fields on this interface are nullable because typically the spec will _not_ have that information available
+ * before the corresponding [ResourceHandler] has resolved the resource.
+ */
+interface ArtifactProvider {
+  val artifactName: String?
+  val artifactType: ArtifactType?
+
+  fun completeArtifactOrNull() =
+    if (artifactName != null && artifactType != null) {
+      CompleteArtifact(artifactName!!, artifactType!!)
+    } else {
+      null
+    }
+}
+
+/**
+ * Implemented by [ResourceSpec] or concrete resource types that (may) contain versioned artifacts, typically compute
+ * resources.
+ */
+interface VersionedArtifactProvider : ArtifactProvider {
+  val artifactVersion: String?
+
+  fun completeVersionedArtifactOrNull() =
+    if (artifactName != null && artifactType != null && artifactVersion != null) {
+      CompleteVersionedArtifact(artifactName!!, artifactType!!, artifactVersion!!)
+    } else {
+      null
+    }
+}
+
+/**
+ * Implemented by [ResourceSpec] or concrete resource types that (may) contain artifact references, typically compute
+ * resources.
+ */
+interface ArtifactReferenceProvider {
+  val artifactReference: String?
+  val artifactType: ArtifactType?
+
+  fun completeArtifactReferenceOrNull() =
+    if (artifactReference != null && artifactType != null) {
+      CompleteArtifactReference(artifactReference!!, artifactType!!)
+    } else {
+      null
+    }
+}
+
+interface ComputeResourceSpec : ResourceSpec, VersionedArtifactProvider, ArtifactReferenceProvider
+
+/**
+ * Simple container of the information defined in [ArtifactProvider] that ensures non-nullability of the fields.
+ */
+data class CompleteArtifact(
```

Comment:
I find these `Complete*` names a bit confusing. What does "complete" mean in this context?

Reply:
That all the properties are non-null, basically. Like I said [here](https://github.com/spinnaker/keel/pull/905#discussion_r395799425), I struggled with this design but was unaware of a better pattern. The interfaces have all nullable fields, but I wanted a type that implements the interface but where the fields are guaranteed to be non-null. Any other way to approach this?